text/2021-12-29-openapi-to-typescript-api-client-and-forms.md

```diff
@@ -0,0 +1,458 @@
+me that **we can write it in the comments**. Yes, like `kubebuilder`'s marker, we
+can also define markers.
+
+So the following are all markers we will using:
+
+- ui:form:enum=action1;action2;action3
+
+  > Indicate what actions we have.
+
+- ui:form:action=action1
+
+  > Indicate this property belongs to which action.
+
+- ui:form:ignore
+
+  > Ignore this property.
+
+For example:
+
+```go
+type AWSChaosSpec struct {
+        // ui:form:enum:ec2-stop,ec2-restart,detach-volume
+        // +kubebuilder:validation:Enum=ec2-stop;ec2-restart;detach-volume
+        Action AWSChaosAction `json:"action"`
+
+        //...
+}
+
+type AWSSelector struct {
+        // Endpoint indicates the endpoint of the aws server. Just used it in test now.
+        // ui:form:ignore
+        // +optional
+        Endpoint *string `json:"endpoint,omitempty"`
+
+        //...
+
+        // DeviceName indicates the name of the device.
+        // ui:form:action:detach-volume
+        // +optional
+        DeviceName *string `json:"deviceName,omitempty" webhook:"AWSDeviceName,nilable"`
+}
+```
+
+The rest of the steps are simple, use regular expressions to read them:
+
+```js
+// Part of the code
+const UI_FORM_ENUM = /ui:form:enum:(.+)\s/;
+
+/**
+ * Get enum array from jsdoc comment.
+ *
+ * @export
+ * @param {string} s
+ * @return {string[]}
+ */
+export function getUIFormEnum(s) {
+  const matched = s.match(UI_FORM_ENUM);
+
+  return matched ? matched[1].split(",") : [];
+}
+
+// ...
+
+// Assuming that the Node is found
+const { escapedText: identifier } = node.name; // identifier
+const comment = node.jsDoc[0].comment;
+
+if (identifier === "action") {
+  // Get all actions
+  actions = getUIFormEnum(comment);
+}
+```
+
+Similarly we can handle other markers in this way. So far we have solved this problem.
+
+#### Shared fields
+
+Now this problem also becomes simple, since we have defined markers, we can default
+to:
+
+**The fields without `ui:form:action:xxx` and `ui:form:ignore` are shared**.
+
+I think this part can be ignored for code details, it is enough to understand this
+given condition.
+
+#### Chaos without action
+
+Here is another case, what if a chaos doesn't have an action? Like `KernelChaos`
+and `TimeChaos`.
+
+This Chaos will output `export default shared` directly.
+
+#### Non-primitive type
+
+If a field is not primitive type, TypeScript will use an `TypeReference` to represent
+it. This is different from primitive types, once the type reference appears, we
+need to recursively get the corresponding type symbol. We will use the checker to
+achieve this:
+
+```ts
+const program = ts.createProgram([source], {
+  target: ts.ScriptTarget.ES2015,
+});
+const sourceFile = program.getSourceFile(source);
+const checker = program.getTypeChecker(); // this is what we need
+
+const type = checker.getTypeAtLocation(typeRef); // get the final type
+```
+
+So we still need to use a new field type to represent it. Here is an example:
+
+```js
```

Comment:
When we update the struct of chaos, do we need to update the js file manually?

Reply:
No, all automatically generated. We only need to care about are these things:

1. Is the field should use the `ui` mark to complete special operations?
2. A meaningful description (comment), the form will use it as a tip.

Reply:
P.S. I think we should also add a new `ui` mark for the custom initial value, like `ui:form:value=1`. This is the one thing I forgot to consider.

Reply:
OK